server/sonar-db-dao/src/main/java/org/sonar/db/measure/MeasureQuery.java

```diff
@@ -0,0 +1,242 @@
+
+  @CheckForNull
+  private final Collection<String> metricKeys;
+
+  @CheckForNull
+  private final Long personId;
+
+  private MeasureQuery(Builder builder) {
+    this(builder.analysisUuid, builder.projectUuids, builder.componentUuids, builder.metricIds, builder.metricKeys, builder.personId);
+  }
+
+  private MeasureQuery(@Nullable String analysisUuid,
+    @Nullable Collection<String> projectUuids,
+    @Nullable Collection<String> componentUuids,
+    @Nullable Collection<Integer> metricIds,
+    @Nullable Collection<String> metricKeys,
+    @Nullable Long personId) {
+    checkArgument(metricIds == null || metricKeys == null, "Metric IDs and keys must not be set both");
+    checkArgument(projectUuids != null || componentUuids != null, "At least one filter on component UUID is expected");
+    checkArgument(componentUuids == null || componentUuids.size() == 1 || (projectUuids != null && projectUuids.size() == 1),
+      "Component UUIDs can only be used when a single project UUID is set");
+
+    this.analysisUuid = analysisUuid;
+    this.projectUuids = projectUuids;
+    this.componentUuids = componentUuids;
+    this.metricIds = metricIds;
+    this.metricKeys = metricKeys;
+    this.personId = personId;
+  }
+
+  public String getAnalysisUuid() {
+    return analysisUuid;
+  }
+
+  @CheckForNull
+  public Collection<String> getProjectUuids() {
+    return projectUuids;
+  }
+
+  @CheckForNull
+  public String getProjectUuid() {
+    return isOnComponents() ? projectUuids.iterator().next() : null;
+  }
+
+  @CheckForNull
+  public Collection<String> getComponentUuids() {
+    return componentUuids;
+  }
+
+  @CheckForNull
+  public String getComponentUuid() {
+    return isOnSingleComponent() ? componentUuids.iterator().next() : null;
+  }
+
+  @CheckForNull
+  public Collection<Integer> getMetricIds() {
+    return metricIds;
+  }
+
+  @CheckForNull
+  public Collection<String> getMetricKeys() {
+    return metricKeys;
+  }
+
+  @CheckForNull
+  public Long getPersonId() {
+    return personId;
+  }
+
+  public boolean returnsEmpty() {
+    return (projectUuids != null && projectUuids.isEmpty())
+      || (componentUuids != null && componentUuids.isEmpty())
+      || (metricIds != null && metricIds.isEmpty())
+      || (metricKeys != null && metricKeys.isEmpty());
+  }
+
+  public boolean isOnProjects() {
+    return projectUuids != null && componentUuids == null;
+  }
+
+  public boolean isOnComponents() {
+    return projectUuids != null && projectUuids.size() == 1 && componentUuids != null;
+  }
+
+  public boolean isOnSingleComponent() {
+    return projectUuids == null && componentUuids != null && componentUuids.size() == 1;
+  }
+
+  @Override
+  public boolean equals(@Nullable Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    MeasureQuery that = (MeasureQuery) o;
+    return Objects.equals(analysisUuid, that.analysisUuid) &&
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Reduce the number of conditional operators (5) used in the expression (maximum allowed 3). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1067)